lib/core/src/Cardano/Wallet.hs

```diff
@@ -1752,6 +1701,180 @@ balanceTransaction
+                    tokenBundleSizeAssessor tl $
+                    pp ^. (#txParameters . #getTokenBundleMaxSize)
+                , certificateDepositAmount =
+                    view #stakeKeyDeposit pp
+                , computeMinimumAdaQuantity =
+                    view #txOutputMinimumAdaQuantity $ constraints tl pp
+                , computeMinimumCost = \skeleton -> mconcat
+                    [ feePadding
+                    , fromCardanoLovelace fee0
+                    , calcMinimumCost tl pp
+                        (defaultTransactionCtx
+                            { txPlutusScriptExecutionCost =
+                                txPlutusScriptExecutionCost })
+                        skeleton
+                    ] `Coin.difference` boringFee
+
+                , computeSelectionLimit = \outputs ->
+                    -- To compute how many inputs we can select, we rely on the
+                    -- existing 'computeSelectionLimit', but subtract the size
+                    -- of the partial tx from 'getTxMaxSize'.
+                    --
+                    -- FIXME: This will double-count inputs and outputs, which
+                    -- could make balancing of large txs fail when they
+                    -- shouldn't.
+                    --
+                    -- We should be able to do something like this instead:
+                    --
+                    -- let constraints = txConstraints pparams tag
+                    --     nWits = estimateNumberOfWitnesses tx
+                    --     budget = getTxMaxSize
+                    --          - serializedSize tx
+                    --          - nWits * (sizeOfWitness constraints)
+                    --  in
+                    --      budget `div` (sizeOfInput constraints)
+                    --
+                    -- Even then we'd like some test coverage. How do we get
+                    -- that?
+                    --
+                    -- forall tx. balanceTx tx == SelecionLimitExceeded
+                    --      ==> forall newTxMaxSize. newTxMaxSize > oldTxMaxSize
+                    --      ==> let tx' = balanceTx {newTxMaxSize} tx
+                    --          tx' == Success ==> size tx' > oldTxMaxSize
+                    --
+                    --  We could also heavily reduce the tx max size and add
+                    --  some goldens...
+                    let
+                        fixedSize = BS.length $ serialisedTx partialTx
+                        tweak (Quantity maxSize) = Quantity
+                            . fromMaybe 0
+                            . intCastMaybe
+                            $ (intCast maxSize) - fixedSize
+                        pp' = over (#txParameters . #getTxMaxSize) tweak pp
+                    in
+                        (view #computeSelectionLimit tl)
+                            pp'
+                            defaultTransactionCtx
+                            outputs
```

Comment:
This is the part I'm uncertain about... Can this be merged as is?

Testing that `balanceTransaction` doesn't create too large txs would require fixing [ADP-1484](https://input-output.atlassian.net/browse/ADP-1484). I would not want to make this PR dependent on that.

Testing that `balanceTransaction` doesn't fail needlessly with `SelectionLimitReached` might be possible as outlined by the comment, but it might take a few days.

Perhaps the
```
let constraints = txConstraints pparams tag
                     --     nWits = estimateNumberOfWitnesses tx
                     --     budget = getTxMaxSize
                     --          - serializedSize tx
                     --          - nWits * (sizeOfWitness constraints)
                     --  in
                     --      budget `div` (sizeOfInput constraints)
```
approach suggested above, but without testing is good enough (for now). Misjudging what we can fit in a tx by a few bytes shouldn't be the end of the world, given that we already may misjudge because of change (ADP-1484)

Reply:
N/A thanks to new approach supported by #3164